pulser/simulation.py

```diff
@@ -388,34 +359,114 @@ def run_noisy(self, initial_state=None,
+            else:
+                if self.basis_name in {'digital', 'all'}:
+                    return 'digital'
+                else:
+                    return 'ground-rydberg'
+
+        def _build_hamiltonian():
             self._extract_samples()
+            self._build_basis_and_op_matrices()
+            if 'initial_state' not in self._config:
+                all_ground = [self.basis['g'] for _ in range(self._size)]
+                self._config['initial_state'] = qutip.tensor(all_ground)
             self._construct_hamiltonian()
-            current_res = self.run(initial_state=initial_state)
-            if self._noise["SPAM"]:
-                total_count += \
-                    current_res.sampling_with_detection_errors(
-                        self.spam_dict, t=t, meas_basis=meas_basis,
-                        N_samples=N_samples_per_run)
-            else:
-                total_count += current_res.sample_state(
-                    t=t, meas_basis=meas_basis, N_samples=N_samples_per_run)
-        prob = Counter({k: v / (N_runs * N_samples_per_run) for
-                        k, v in total_count.items()})
-        return NoisyResults(prob, self._size, self.basis_name,
-                            meas_basis=meas_basis)
+
+        def _run_solver(t_list=None, as_subroutine=False,
+                        measurement_basis=None):
+            if not as_subroutine:
+                # CLEAN SIMULATION:
+                # Build (clean) hamiltonian
+                _build_hamiltonian()
+                measurement_basis = _assign_meas_basis()
+
+            time_list = t_list if t_list else self._times
+            result = qutip.mesolve(self._hamiltonian,
+                                   self._config['initial_state'],
+                                   time_list,
+                                   c_ops=[],
+                                   progress_bar=progress_bar,
+                                   options=qutip.Options(max_step=5,
+                                                         **options)
+                                   )
+            return CleanResults(result.states, self.dim, self._size,
+                                self.basis_name, measurement_basis)
+
+        if self._noise:
+            #  NOISY SIMULATION:
+            # We run the system multiple times
+            total_count = Counter()
+            for _ in range(self._config['runs']):
+                # At each run, new random noise
+                self._prepare_spam_detune()
+                _build_hamiltonian()
+                meas_basis = _assign_meas_basis()
+                # Get CleanResults instance from sequence with added noise:
+                res_with_noise = _run_solver(as_subroutine=True,
+                                             measurement_basis=meas_basis)
+
+                # Extract statistics at eval time (final time by default):
+                if self._noise['SPAM']:
```

Comment:
Doesn't work if there is no 'SPAM' noise in the dictionary : hence why I think we should change it to a list and write
`if 'SPAM' in self._noise` instead.